HDFOutputer.h

```diff
@@ -0,0 +1,65 @@
+#if !defined(HDFOutputer_h)
+#define HDFOutputer_h
+
+
+#include <vector>
+#include <string>
+#include <cstdint>
+#include <fstream>
+
+
+#include "OutputerBase.h"
+#include "EventIdentifier.h"
+#include "SerializerWrapper.h"
+#include "DataProductRetriever.h"
+
+#include "SerialTaskQueue.h"
+
+#include "HDFCxx.h"
+
+using product_t = std::vector<char>;
+
+namespace cce::tf {
+  class HDFOutputer : public OutputerBase {
+    public:
+    HDFOutputer(std::string const& iFileName, unsigned int iNLanes);
+    HDFOutputer(HDFOutputer&&) = default;
+    HDFOutputer(HDFOutputer const&) = default;
+    ~HDFOutputer();
+
+  void setupForLane(unsigned int iLaneIndex, std::vector<DataProductRetriever> const& iDPs) final;
+
+  void productReadyAsync(unsigned int iLaneIndex, DataProductRetriever const& iDataProduct, TaskHolder iCallback) const final;
+  bool usesProductReadyAsync() const final {return true;}
+
+  void outputAsync(unsigned int iLaneIndex, EventIdentifier const& iEventID, TaskHolder iCallback) const final;
+  
+  void printSummary() const final;
+
+ private:
+  static inline size_t bytesToWords(size_t nBytes) {
+    return nBytes/4 + ( (nBytes % 4) == 0 ? 0 : 1);
+  }
+
+  void output(EventIdentifier const& iEventID, std::vector<SerializerWrapper> const& iSerializers);
+  void writeFileHeader(EventIdentifier const& iEventID, std::vector<SerializerWrapper> const& iSerializers);
+
+  std::vector<std::vector<char>> writeDataProductsToOutputBuffer(std::vector<SerializerWrapper> const& iSerializers) const;
+std::pair<product_t, std::vector<size_t>> get_prods_and_sizes(std::vector<product_t> & input,int prod_index,int stride);
+private:
+  File file_;
+  mutable SerialTaskQueue queue_;
+  std::vector<std::pair<std::string, uint32_t>> dataProductIndices_;
+  mutable std::vector<std::vector<SerializerWrapper>> serializers_;
+  bool firstTime_ = true;
+  
+  std::vector<product_t> products_; 
+  std::vector<size_t> offsets_ = decltype(offsets_)(1000, 0);
+  mutable std::atomic<int> batch_ = 0;
```

Comment:
batch_ is keeping track of total number of events to accumulate before a file write happens.

Reply:
`batch_` is only read and modified in `output()` and that function is called serially via the `SerialTaskQueue`. Therefore there is no need for it to be atomic since the `SerialTaskQueue` already guarantees single access.
(Plus the way you are using `batch_` in `output()` wouldn't do what you wanted it to if `output()` was called via multiple threads concurrently).

Reply:
I see, okay.